docs/dev/metrics.md

```diff
@@ -0,0 +1,108 @@
+
+2. **Avoid High-Cardinality Attributes**: Avoid using high-cardinality labels (e.g., user IDs, request IDs) as they can significantly impact storage, query performance, and resource usage.
+   - **Why**: Managing cardinality helps control resource usage and maintain cost-effective metrics, aligning with the Developer Experience priority by reducing performance bottlenecks.
+
+3. **Use lowercase snake_case for Naming**: Follow `snake_case` for metric names, aligning with widely adopted conventions in Prometheus.
+   - **Why**: Adhering to common casing conventions improves readability, reduces errors, and increases developer familiarity, which enhances Developer Experience.
+
+4. **Non-Pluralized Naming**: Avoid plural forms in domains and metric names for uniformity, unless the value being recorded represents discrete instances of a [countable quantity](https://wikipedia.org/wiki/Count_noun). Generally, the name is pluralized only if the unit of the metric in question is a non-unit, e.g.,  `flight_requests`, `results_cache_hits`
+   - **Why**: Singular terms are simpler and prevent naming inconsistencies, improving readability and maintaining consistency across metrics.
+
+5. **Begin with a Domain Prefix**: Start each metric with a domain prefix or namespace, especially for protocol-based metrics.
+   - **Why**: Domain prefixes provide immediate context, allowing developers to quickly understand a metric’s scope and relevance, supporting both Clarity and Developer Experience.
+
+6. **Avoid Redundancy in Naming**: Avoid repeating terms that are implied by the context or domain, making metric names shorter and more readable. Example: `results_cache_hits` instead of `results_cache_cache_hits`, `query_duration_ms` instead of `query_query_duration_ms`.
+   - **Why**: Redundancy adds unnecessary complexity and reduces readability. Shorter, more focused names enhance understanding and ease of use (Developer Experience principle).
+
+7. **Counter Naming Conventions**: Avoid `_total` suffix in counter names e.g., `flight_requests` instead of `flight_requests_total`.   - **Why**: Standardized counter naming aligns with Prometheus conventions, reducing confusion and supporting Developer Experience by avoiding unnecessary naming variation.
+
+8. **UpDownCounter Naming**: Use `_count` suffix and pluralization only when recording discrete instances of countable items, e.g., `db_client_active_connections_count`.
+   - **Why**: Clearly distinguishing counter-based metrics improves interpretability, which enhances the Developer Experience by making data analysis more intuitive.
+
+9. **Include Units in Name for Unit-Based Metrics**: Specify the unit (e.g., `_ms` for milliseconds, `_bytes` for data size) within metric names where applicable. Units should adhere to UCUM standards for consistent scientific notation.
```

Comment:
Might be worth adding a note that this goes against the OTel recommendation, but we are choosing to go against that for prometheus metric compatibility.

Reply:
Also UCUM states that `_by` should be the suffix for bytes, not `_bytes` - is that an explicit decision?

Reply:
`_bytes` suffix proposed based on `Prioritize Developer Experience (DX)` principle as `_by` is not very intuitive to use in metric name.

Note: `with_unit("By")` is proposed to be used in the same time. While not very consistent, it should provide a good developer experience (DX) when using the metric and ensure the metric unit adheres to the UCUM standard.